### PR TITLE
Bug 1184077 - [NGA][Contacts] VCARD loader as independent page

### PR DESCRIPTION
**Pull request by mancas**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9da06efc51e1511dd4613b68844c38d11bed7238

File: apps/communications/contacts/views/load/js/load_ui.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    header = document.getElementById('multiple-select-view-header');
+    saveButton = document.getElementById('save-button');
+    container = document.getElementById('multiple-select-container');
+    template = document.getElementById('contactToImportTemplate');
+    title = document.getElementById('multiple-select-view-title');
+
+    saveButton.addEventListener('click',
+      dispatchEvent.bind(null, 'saveAction',
+                        {'contactsToImport': contactsToImport}));
+    header.addEventListener('action', dispatchEvent.bind(null, 'closeAction'));
```
Comment (borjasalguero) [inline apps/communications/contacts/views/load/js/load_ui.js]: Could you split this in `cacheElements()` and `addListeners()`?

File: apps/communications/contacts/views/load/js/load_ui.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  function dispatchEvent(name, data) {
+    window.dispatchEvent(new CustomEvent(name, {'detail': data}));
+  }
+
+  exports.LoadUI = {
+    'init': init,
+    'render': render
+  };
+})(window);
```
Comment (borjasalguero) [inline apps/communications/contacts/views/load/js/load_ui.js]: One more line in here! :)

File: apps/communications/contacts/test/unit/views/load/boot_test.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+require('/shared/js/lazy_loader.js');
+requireApp('communications/contacts/views/load/js/boot.js');
+requireApp('communications/contacts/views/load/js/vcard_handler.js');
+requireApp('communications/contacts/test/unit/mock_contacts_list.js');
+
+
+suite('Open VCards Boot', function() {
+  var stub;
+  suiteSetup(function() {
+    stub = sinon.stub(LazyLoader, 'load', function(files) {
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/boot_test.js]: nit: more descriptive name (*e.g. lazyLoad_stub*)

File: apps/communications/contacts/views/load/load.html [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    <link href="/shared/style/edit_mode.css" rel="stylesheet">
+    <link href="/shared/style/animations.css" rel="stylesheet">
+
+    <link href="/contacts/style/app.css" rel="stylesheet">
+    <link href="/contacts/style/contacts.css" rel="stylesheet">
+    <link href="/contacts/style/overlay.css" rel="stylesheet">
+    <link href="/shared/pages/import/style/search.css" rel="stylesheet">
+    <link href="/shared/pages/import/style/curtain_frame.css" rel="stylesheet">
+    <link href="/contacts/style/status.css" rel="stylesheet">
+    <link href="/contacts/style/fb_extensions.css" rel="stylesheet">
```
Comment (borjasalguero) [inline apps/communications/contacts/views/load/load.html]: Do we need this?

File: apps/communications/contacts/views/load/load.html [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    <script defer src="/shared/elements/gaia_subheader/script.js"></script>
+    <script defer src="/shared/elements/gaia-header/dist/gaia-header.js"></script>
+
+
+    <!-- Localization -->
+    <link rel="localization" href="/shared/elements/gaia_menu/locales/gaia_menu.{locale}.properties">
+
+    <script defer src="/shared/js/html_imports.js"></script>
+    <script defer src="/shared/js/lazy_loader.js"></script>
+    <script defer src="/contacts/js/utilities/performance_helper.js"></script>
```
Comment (borjasalguero) [inline apps/communications/contacts/views/load/load.html]: Do we need this here?

File: apps/communications/contacts/views/load/load.html [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    <meta name="defaultLanguage" content="en-US">
+    <meta name="availableLanguages" content="en-US">
+    <link rel="localization" href="/contacts/locales/contacts.{locale}.properties">
+    <link rel="localization" href="/shared/locales/import_contacts/import_contacts.{locale}.properties">
+    <link rel="localization" href="/shared/locales/date/date.{locale}.properties">
+    <link rel="localization" href="/shared/locales/phone_types/phone_types.{locale}.properties">
+
+
+    <!-- Some CSS magic needed to work in a separate html -->
+    <link href="/contacts/views/form/style/overwrite.css" rel="stylesheet">
```
Comment (borjasalguero) [inline apps/communications/contacts/views/load/load.html]: Copy&Paste in here! ;)

File: apps/communications/contacts/test/unit/views/load/load_ui_test.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      var contactList = document.getElementById('multiple-select-container');
+      var listItems = contactList.querySelectorAll('li');
+      assert.equal(listItems.length, 3);
+
+      assert.equal(listItems[0].textContent.trim(), 'Pepito A\n        Test');
+      assert.equal(listItems[1].textContent.trim(), 'Pepito BA\n        Test');
+      assert.equal(listItems[2].textContent.trim(), 'Antonio CC\n        Test');
+    });
+  });
+});
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/views/load/load_ui_test.js]: One more line!

File: apps/communications/contacts/test/unit/views/load/load_controller_test.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
+/* global LazyLoader, ContactsService, MockMatcher, MockContactsList */
+/* global LoadController, utils */
+
+'use strict';
+
+require('/shared/js/contacts/import/utilities/status.js');
+requireApp('communications/contacts/test/unit/mock_contacts_match.js');
+requireApp('communications/contacts/services/contacts.js');
+requireApp('communications/contacts/test/unit/mock_contacts_list.js');
+require('/shared/js/lazy_loader.js');
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/load_controller_test.js]: nit: all `/shared/`resources together

File: apps/communications/contacts/test/unit/views/load/load_controller_test.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        type:'text/x-vcard'
+      }
+    },
+    postResult: function foo() {}
+  };
+
+  suiteSetup(function() {
+    LoadController.init();
+    LoadController.setActivity(fakeActivity);
+    stubLazyLoader = sinon.stub(LazyLoader, 'load', function(files, callback) {
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/load_controller_test.js]: if you use `this.sinon.stub` on `setup` instead, then there's no need of restoring it later.

File: apps/communications/contacts/test/unit/views/load/load_controller_test.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+  setup(function() {
+    contacts = MockContactsList();
+    stubContactServiceSave = this.sinon.stub(ContactsService, 'save',
+                                              function(contact, onerror) {
+                                                onerror(null);
+                                              });
+    stubUtilsStatus = this.sinon.stub(utils.status, 'show', function show() {});
+
+    realMatcher = window.Matcher;
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/load_controller_test.js]: why not put this on `suiteSetup`?

File: apps/communications/contacts/test/unit/views/load/load_controller_test.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+      window.dispatchEvent(new CustomEvent('closeAction'));
+    });
+  });
+
+  suite('Import all button', function() {
+    test(' > must import all contacts received', function(done) {
+      var spyMatcher = this.sinon.spy(window.Matcher, 'match');
+      var spyPostResult = this.sinon.spy(fakeActivity, 'postResult');
+      var clock = sinon.useFakeTimers();
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/load_controller_test.js]: nit: all `var` declarations on the `setup`
`this.sinon.useFakeTimers`

File: apps/communications/contacts/test/unit/views/load/load_ui_test.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+suite('LoadUI', function() {
+
+  var realMozL10n, contacts;
+
+  suiteSetup(function(done) {
+    realMozL10n = navigator.mozL10n;
+    navigator.mozL10n = MockL10n;
+
+    // Load HTML
+    loadBodyHTML('/contacts/views/load/load.html');
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/load_ui_test.js]: better to do it on `setup` to clean it before every test

File: apps/communications/contacts/test/unit/views/load/vcard_handler_test.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  teardown(function() {
+    window.VCardReader = realVCardReader;
+    realVCardReader = null;
+  });
+
+  suite('handle', function() {
+    function initializeVCardReader(filename) {
+      return new Promise(function(resolve, reject) {
+        var req = new XMLHttpRequest();
+        var vCardUrl = '/apps/communications/contacts/test/unit/vcards/' +
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/vcard_handler_test.js]: nit: make the string a const so we can reuse

File: apps/communications/contacts/test/unit/views/load/vcard_handler_test.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+          done();
+        });
+      });
+    });
+  });
+
+  suite('getFileName', function() {
+    test(' > vCard filename', function() {
+      var name = VCardHandler.getFileName(
+        '/apps/communications/contacts/test/unit/vcards/vcard_21_multiple.vcf');
```
Comment (fcampo) [inline apps/communications/contacts/test/unit/views/load/vcard_handler_test.js]: nit: const to reuse

File: apps/communications/contacts/views/load/js/boot.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+ * Once localization and all the basic JS/CSS/Elements are loaded,
+ * we will initialize UI and Controller. Both JS classes *must* be
+ * independent and will communicate through events.
+ */
+
+window.addEventListener('DOMContentLoaded', function() {
+  LazyLoader.load(['/shared/js/l10n.js']).then(function() {
+    LazyLoader.load([
+      document.getElementById('multiple-select-view')
+    ]).then(function() {
```
Comment (fcampo) [inline apps/communications/contacts/views/load/js/boot.js]: no need for the `.then` if we don't have code to fill with

File: apps/communications/contacts/views/load/js/load_controller.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+              success: () => {
+                numDupsMerged++;
+                doContinue(true);
+              },
+              error: doContinue
+            };
+            contacts.adaptAndMerge(contact, matches, callbacks);
+          },
+          onmismatch: () => {
+
```
Comment (fcampo) [inline apps/communications/contacts/views/load/js/load_controller.js]: nit: extra line

File: apps/communications/contacts/views/load/js/load_controller.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    });
+
+    function doContinue(isContactImported) {
+      parsedContacts++;
+      if (isContactImported) {
+        importedContacts++;
+      }
+
+      if (contactsToImport.length <= parsedContacts) {
+        utils.status.show({
```
Comment (fcampo) [inline apps/communications/contacts/views/load/js/load_controller.js]: `status.show` is a visual change, it should be dealt with on the UI, not the controller.
maybe sending an event that _ui catches?

File: apps/communications/contacts/test/unit/views/load/load_ui_test.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        done();
+      }
+    );
+  });
+
+  suiteTeardown(function() {
+    navigator.mozL10n = realMozL10n;
+    realMozL10n = null;
+    window.LoadUI = null;
+    document.body.innerHTML = '';
```
Comment (arcturus) [inline apps/communications/contacts/test/unit/views/load/load_ui_test.js]: Following previous comment from Fernando, this should go in the teardown

**Comment by arcturus** [comment]
The name of the view `load` is confusing to me. If this is related to vcard, i will prefix it to avoid any confusion.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7496b7bc61476fb83a559bc07e6ca734dab806db

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9011de4fa66aef67dc4d8ebdbddcb8ec262663d1

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9011de4fa66aef67dc4d8ebdbddcb8ec262663d1

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=292f081d8879adcc86cee14880e4fee3625586cf

**Pull request merged**